### PR TITLE
feat(bulk): merge history capable south initialization

### PR DESCRIPTION
**Pull request by kukukk**

### CODE REVIEW

File: src/south/ProtocolHandler.class.js [hunk 1]
```diff
@@ -75,7 +108,24 @@ class ProtocolHandler {
...
+      // Disable ESLint check because we need for..of loop to support async calls
+      // eslint-disable-next-line no-restricted-syntax
+      for (const scanMode of Object.keys(this.lastCompletedAt)) {
+        // Disable ESLint check because we want to get the values one by one to avoid parallel access to the SQLite database
+        // eslint-disable-next-line no-await-in-loop
+        const lastCompletedAt = await this.getConfig(`lastCompletedAt-${scanMode}`)
+        this.lastCompletedAt[scanMode] = lastCompletedAt ? new Date(lastCompletedAt) : defaultLastCompletedAt
+        this.logger.info(`Initializing lastCompletedAt for ${scanMode} with ${this.lastCompletedAt[scanMode]}`)
+      }
+    }
```
Comment (burgerni10) [inline src/south/ProtocolHandler.class.js:127]: Shouldn't this block be in the constructor instead of the connect ? In the if block l.74

**Review by burgerni10 — CHANGES_REQUESTED** [review]

File: src/south/ProtocolHandler.class.js (hunk 1 above)
Comment (kukukk) [reply src/south/ProtocolHandler.class.js:127]: It is not possible to use await in constructor.
(Using await requires the function to be async. If the function is async it returns a Promise. The constructor returns the newly created object)

File: src/south/ProtocolHandler.class.js [hunk 2]
```diff
@@ -64,6 +67,36 @@ class ProtocolHandler {
...
+              .map((point) => point.nodeId)
+            this.lastCompletedAt[scanGroup.scanMode] = new Date().getTime()
+            this.ongoingReads[scanGroup.scanMode] = false
+            return {
+              name: scanGroup.scanMode,
+              ...scanGroup,
+              points,
+            }
+          })
+        } else {
```
[the remark this answers is not in the document]
Comment (kukukk) [reply src/south/ProtocolHandler.class.js:95]: There is a main if, so this logic applies only to history capable South. If it is a history capable South we check whether it has scanMode defined or it contains scanGroups definition. This else will be triggered only if there is no scanMode nor scanGroups. Similar logic is used in OIBusEngine.class.js line 202.

File: src/south/ProtocolHandler.class.js [hunk 3]
```diff
@@ -54,7 +54,10 @@ class ProtocolHandler {
     this.logger.changeParameters(this.engineConfig.logParameters, logParameters)
 
     if (this.supportedModes) {
-      const { supportListen, supportLastPoint, supportFile } = this.supportedModes
+      const { supportListen, supportLastPoint, supportFile, supportHistory } = this.supportedModes
+      if (!supportListen && !supportLastPoint && !supportFile && !supportHistory) {
+        this.logger.error(`${this.constructor.name} should support at least 1 operation mode.`)
```
[the remark this answers is not in the document]
Comment (kukukk) [reply src/south/ProtocolHandler.class.js:59]: This peace of code only checks whether the South implementation properly calls super(). If any of these ifs are true, it means that the code itself is wrong. So these logs are more for the developer and not for the user, so knowing the problematic class seems to me more useful in this case.

**Review by burgerni10 — APPROVED** [review]

**Pull request merged**